mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,516 @@
+                   'sigma_select' and/or 'sigma_replace' both equal to some
+                   float.
+        .
+    """
+    # set the seed for sampling, for replacement and selection:
+    (random_state_select, random_state_replace) = random_state.randint(
+        10 ** (8), size=2
+    )
+    random_state_select = np.random.RandomState(random_state_select)
+    random_state_replace = np.random.RandomState(random_state_replace)
+    # get the operations from the circuit and find the non-cliff angles:
+    operations = np.array(list(circuit.all_operations()))
+    positions = np.linspace(1, len(operations), len(operations))
+    gates = _get_gates(operations)
+    mask = np.array(
+        [isinstance(i, cirq.ops.common_gates.ZPowGate) for i in gates]
+    )
+    r_z_gates = operations[mask]
+    r_z_positions = positions[mask]
+    angles = _get_arguments(r_z_gates)
+    mask_non_cliff = ~_is_clifford_angle(angles)
+    rz_non_cliff = angles[mask_non_cliff]
+    pos_non_cliff = r_z_positions[mask_non_cliff]
+    rz_non_cliff_copy = rz_non_cliff.copy()
+    sigma_select = additional_options.setdefault("sigma_select", 0.5)
+    sigma_replace = additional_options.setdefault("sigma_replace", 0.5)
+    if (
+        "sigma_select" not in additional_options
+        and "sigma_replace" not in additional_options
+    ):
+        raise Exception(
+            "additional options must be dicitonary with \
+                        keys containing one or both of "
+            '"sigma_select" and "sigma_replace" both \
+                        equal to some positive float'
+        )
+    # Choose non Clifford gates to change according to selection methods:
+    columns_to_change = _select(
+        rz_non_cliff_copy,
+        fraction_non_clifford,
+        method_select,
+        sigma_select,
+        random_state_select,
+    )
+    rz_non_cliff_selected = rz_non_cliff_copy[columns_to_change]
+    pos_selected = pos_non_cliff[columns_to_change]
+    # Now the non Clifford gates have been selected, we need to decide which
+    # Clifford gate to replace them with.
+    # to store original angles replaced:
+    angles_original = rz_non_cliff_selected.copy()
+    rz_non_cliff_selected = _replace(
+        rz_non_cliff_selected,
+        method_replace,
+        sigma_select,
+        sigma_replace,
+        random_state_replace,
+    )
+    # to store replaced angles:
+    angles_replaced = rz_non_cliff_selected.copy()
+    # build projected cirucit:
+    projected_circuit = circuit.copy()[0:0]
+    count = 0
+    for o, op in enumerate(operations):
+        if (o + 1) in pos_selected:
+            qubit = op.qubits[0]
+            parameter = rz_non_cliff_selected[count]
+            operation = cirq.ops.rz(parameter)
+            projected_circuit.append(operation(qubit))
+            count += 1
+        else:
+            projected_circuit.append(op)
+    return projected_circuit, angles_original, angles_replaced
```

Comment:
Is it just for debugging to return the original and replaced angles? If so it should be removed.